### PR TITLE
Map loading and Transform/Odom topic refactoring 

### PR DESCRIPTION
**Pull request by TirelessDev**
This PR covers two changes to the hdl_graph_slam master. 

- Introduces an initial first pass of tying together the loading functionality of g2o and graph_slam within a service request endpoint. 
- Refactors the code slightly so that published topic frames can be easier defined in the launch files.

To the first point. We have added a `load` service endpoint that takes a string of the path to a map previously dumped using the `dump` endpoint. This is a folder that contains pose graph information and keyframes.
```
{
  "path": "/tmp/dumped/map"
}
```
image

This loading should be performed on a fresh start of hdl_graph_slam, and before new point clouds are received. After loading new point clouds will be used to extend the loaded map. 

Of particular importance to know is that the current odometry information from the previous map generation process is not retained, ie where it figured the Velodyne frame was within the Map frame. This means that after starting a new instance of hdl_graph_slam and after loading a previous map the "current estimated position" of the Velodyne will be at the origin. This is super important because if point clouds don't then arrive and describe the same local environment, the registration to the existing map will be incorrect. There may be ways to help alleviate this problem, perhaps by providing an initial position estimate to update the internal odom estimate. However, this is a topic for future work.

An example of the std output from the loading service call. 
```
loading data from:/tmp/dumped/map
loading pose graph...
nodes  : 11
edges  : 10
kernels: 0
loaded 9 keyframes
loaded special nodes - anchor_node: 1 anchor_edge: -2 floor_node: 8
snapshot updated
loading successful
```

With respect to the second lot of changes, to the topic subscriptions. These basically involved some minor modifications to the nodelet code to allow additional parameters to be passed into the node to configure the transform frame names. Basically, just to remove hardcoded values like `"/odom"` so we can avoid conflicts when integrating with other systems on the robot.

### CODE REVIEW

**Comment by koide3** [comment]
Thanks a lot for the nice PR. I pushed a fix for a problem on the CI that wrongly failed. Could you merge ```master``` branch to re-run the CI?

The PR looks nice, and I'm going to merge it with slight modifications. But, please note that my review would be slow because I'm a bit busy at this moment for writing a paper...

File: src/hdl_graph_slam/map2odom_publisher.py [hunk 1]
```diff
@@ -6,7 +6,9 @@
 
 
 class Map2OdomPublisher:
-	def __init__(self):
+	def __init__(self, odom_frame_id = 'slam_odom', map_frame_id = 'slam_map'):
```
Comment (koide3) [inline src/hdl_graph_slam/map2odom_publisher.py:9]: Use "odom" and "map"

File: src/hdl_graph_slam/map2odom_publisher.py [hunk 2]
```diff
@@ -30,7 +32,12 @@ def spin(self):
 
 def main():
 	rospy.init_node('map2odom_publisher')
-	node = Map2OdomPublisher()
+
+	# get some parameters to define what default frame_id's should be used while we wait for our first odom message
+	map_frame_id = rospy.get_param('~map_frame_id', 'slam_map')
```
Comment (koide3) [inline src/hdl_graph_slam/map2odom_publisher.py:37]: 'map' and 'odom'

**Review by koide3 — COMMENTED** [review]
Sorry for the very late response. The PR looks fine, but I would like to request to revert a few points to keep the default behavior unchanged because some changes can affect the systems of existing users.

**Comment by marcelinomalmeidan** [comment]
@koide3, I might be interested in these changes, so I can wrap up the work of merging this PR. Is this something that still interests you?
I was looking at the code in here, and I am wondering if you have any reasoning of what the `anchor_edge` means when we save or load a map. It seems that every edge that we create has an `id = -2` (I added prints to my code), so it doesn't seem to be relevant for anything. Would you have a different interpretation for it?

**Comment by TirelessDev** [comment]
Hey guys, apologies for neglecting this. I moved into a different role and didn't have time to continue working on it / totally forgot. My bad :). 

Though it does work, this PR was more a PoC than a serious attempt at getting it working robustly. My understanding at the time was somewhat surface level, so there is certainly more that can be done to get map reuse working to a level where it would be useful in practice. From memory, some helper processes to aid in roughly aligning the map on load would go a long way. The kidnapped robot problem is still real. 

I had renamed the topics to add clarity for our use context and remove contentions with other systems we were running ( We used this on a Boston Dynamics Spot in conjunction with a number of other solutions ). However, I agree it makes sense to change back.

I don't have the capability to test this anymore, but I can make the requested changes and push a new commit to the PR.

File: apps/hdl_graph_slam_nodelet.cpp [hunk 3]
```diff
@@ -75,6 +76,7 @@ class HdlGraphSlamNodelet : public nodelet::Nodelet {
     private_nh = getPrivateNodeHandle();
 
     // init parameters
+    published_odom_topic = private_nh.param<std::string>("published_odom_topic", "/hdl_graph_slam/odom");
```
[the remark this answers is not in the document]
Comment (TirelessDev) [reply apps/hdl_graph_slam_nodelet.cpp]: Reverted

File: apps/hdl_graph_slam_nodelet.cpp [hunk 4]
```diff
@@ -802,6 +808,95 @@ class HdlGraphSlamNodelet : public nodelet::Nodelet {
...
+
+    std::string directory = req.path;
+
+    std::cout << "loading data from:" << directory << std::endl;
+
+    // Load graph.
+    graph_slam->load(directory + "/graph.g2o");
+    
+    // Load keyframes by looping through key frame indexes that we expect to see.
+    for(int i = 0; i < 1000; i++) { // un magic this max iterator
```
[the remark this answers is not in the document]
Comment (TirelessDev) [reply apps/hdl_graph_slam_nodelet.cpp]: Have added a zero, still feels dirty. Is there a better way of handling "magic" numbers like this?

File: apps/scan_matching_odometry_nodelet.cpp [hunk 5]
```diff
@@ -64,6 +64,7 @@ class ScanMatchingOdometryNodelet : public nodelet::Nodelet {
    */
   void initialize_params() {
     auto& pnh = private_nh;
+    published_odom_topic = private_nh.param<std::string>("published_odom_topic", "/hdl_graph_slam/odom");
```
[the remark this answers is not in the document]
Comment (TirelessDev) [reply apps/scan_matching_odometry_nodelet.cpp]: Done

File: launch/hdl_graph_slam.launch [hunk 6]
```diff
@@ -1,40 +1,52 @@
 <?xml version="1.0"?>
 <launch>
   <!-- arguments -->
-  <arg name="nodelet_manager" default="velodyne_nodelet_manager" />
-  <arg name="enable_floor_detection" default="false" />
-  <arg name="enable_gps" default="false" />
+  <arg name="nodelet_manager" default="hdl_graph_slam_nodelet_manager" />
```
[the remark this answers is not in the document]
Comment (TirelessDev) [reply launch/hdl_graph_slam.launch]: Reverted

File: launch/hdl_graph_slam.launch [hunk 7]
```diff
@@ -1,40 +1,52 @@
...
+  <!-- topic for hdl to find velodyne points -->
   <arg name="points_topic" default="/velodyne_points" />
-  <arg name="map_frame_id" default="map" />
-  <arg name="lidar_odom_frame_id" default="odom" />
+  <!-- what should the map frame be called -->
+  <arg name="map_frame_id" default="slam_map" />
+  <!-- what should the base link frame be called -->
+  <arg name="base_link_frame_id" default="slam_base_link" />
+  <!-- what should the odom frame be called -->
+  <arg name="lidar_odom_frame_id" default="slam_odom" />
```
[the remark this answers is not in the document]
Comment (TirelessDev) [reply launch/hdl_graph_slam.launch]: Reverted

File: launch/hdl_graph_slam.launch [hunk 8]
```diff
@@ -1,40 +1,52 @@
...
+
+  <!-- hdl_graph_slam will also publish an odom message -->
+  <!-- what should the topic be for odom messages published by hdl-graph-slam -->
+  <arg name="published_odom_topic" default="/hdl_graph_slam/odom" />
```
[the remark this answers is not in the document]
Comment (TirelessDev) [reply launch/hdl_graph_slam.launch]: Reverted

File: src/hdl_graph_slam/map2odom_publisher.py (hunk 1 above)
Comment (TirelessDev) [reply src/hdl_graph_slam/map2odom_publisher.py]: Done

File: src/hdl_graph_slam/map2odom_publisher.py (hunk 2 above)
Comment (TirelessDev) [reply src/hdl_graph_slam/map2odom_publisher.py]: Done

**Comment by marcelinomalmeidan** [comment]
Hey @TirelessDev, thanks for coming back and making some new commits. Let me know when you're done with new commits, I can test them for you with some test bags.
And you did clarify the question on my mind, as this isn't solving the kidnapped robot problem yet. Still, it is a good step in the right direction. I will see if I can find some time to work on that once this PR finally merges

**Comment by TirelessDev** [comment]
@marcelinomalmeidan Not a prob; it should be up to date with master now and have those default names reverted. 

Adding the extra `/hdl_graph_slam` prefix to the topics was simply a compatibility thing. I was testing numerous slam solutions at the same time. HDL graph slam was my favourite :)

Yeah, as mentioned in the PR, the robot needs to be in a very similar position and orientation when booting up for the map to converge correctly. Practically this wasn't an issue for us, as Spot always returned to the same spot to charge. However, one of the interesting drivers for map reuse for us was actually map sharing and map collaboration. While finding kidnapped robots in SLAM maps is an entire research topic, there are practical ways that the problem can be worked around. Having solid anchors ( QR codes ) in the map to help localise and provide an initial pose correction would probably be the easiest. However, the map as it stands would need to be extended to support this sort of anchor. Alternatively, such behaviour could be sideloaded to a different service quite easily. 

So much potential, so little time.

File: apps/hdl_graph_slam_nodelet.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (marcelinomalmeidan) [reply apps/hdl_graph_slam_nodelet.cpp]: Maybe list the number of files/folders in the given directory, then for sure this "magic number" will be less than or equal to the number or items in the directory

[the remark this answers is not in the document]
Comment (TirelessDev) [reply apps/hdl_graph_slam_nodelet.cpp]: Good idea, see if what I have just added builds.

**Review by marcelinomalmeidan — COMMENTED** [review]
The new solution worked to me! It seems good to go, IMO
Unfortunately, I can't approve this PR, but hopefully we'll hear soon from @koide3

**Comment by koide3** [comment]
Wow, thanks guys for your work!  I'll make time to review and approve the PR in a few days.

File: apps/hdl_graph_slam_nodelet.cpp [hunk 9]
```diff
@@ -802,6 +808,103 @@ class HdlGraphSlamNodelet : public nodelet::Nodelet {
...
+
+    std::cout << "loading data from:" << directory << std::endl;
+
+    // Load graph.
+    graph_slam->load(directory + "/graph.g2o");
+    
+    // Iterate over the items in this directory and count how many sub directories there are. 
+    // This will give an upper limit on how many keyframe indexes we can expect to find.
+    boost::filesystem::directory_iterator begin(directory), end;
+    int maxDirectoryCount = std::count_if(begin, end,
```
[the remark this answers is not in the document]
Comment (TirelessDev) [reply apps/hdl_graph_slam_nodelet.cpp]: Good suggestion, have updated now

File: apps/hdl_graph_slam_nodelet.cpp [hunk 10]
```diff
@@ -802,6 +808,103 @@ class HdlGraphSlamNodelet : public nodelet::Nodelet {
...
+            }
+          } 
+        }
+      } else if(token == "floor_node") {
+        int id = 0;
+        ifs >> id;
+        floor_plane_node = static_cast<g2o::VertexPlane*>(graph_slam->graph->vertex(id));
+      }
+    }
+    std::cout << "loaded special nodes - anchor_node: "<< anchor_node->id() << " anchor_edge: "<< anchor_edge->id() << " floor_node: "<< floor_plane_node->id() << std::endl;
```
Comment (marcelinomalmeidan) [inline apps/hdl_graph_slam_nodelet.cpp:888]: there is a minor issue in this line. If either `anchor_node`, `anchor_edge` or `floor_plane_node` is a null pointer, then we cannot print `->id()` for them.
I'd suggest only printing when they're not null pointers. I think that only `anchor_node` needs to be non-null to be able to load the graph successfully

**Comment by koide3** [comment]
The code now looks good to me, and I did a quick test in my environment . I'll merge this PR once the last issue pointed by @marcelinomalmeidan is fixed.

**Comment by marcelinomalmeidan** [comment]
@TirelessDev ping :-)

**Comment by TirelessDev** [comment]
Not a problem. Sorry for missing that final semicolon. There is always something, looks like you caught it in the merge though.

**Comment by marcelinomalmeidan** [comment]
@TirelessDev thanks for finalizing the PR! 
@koide3 you're very welcome!

**Pull request merged**